pyproject.toml

```diff
@@ -0,0 +1,146 @@
+[tool.poetry]
+name = "prompting"
+version = "2.5"
+description = "Subnetwork 1 runs on Bittensor and is maintained by Macrocosmos. It's an effort to create decentralised AI"
+authors = ["Kalei Brady, Dmytro Bobrenko, Felix Quinque, Steffen Cruz"]
+readme = "README.md"
+
+[tool.black]
+line-length = 130
+target-version = ['py310']
+include = '\.pyi?$'
+exclude = '''
+/(
+    \.git
+| \.hg
+| \.mypy_cache
+| \.tox
+| \.venv
+| _build
+| buck-out
+| build
+)/
+'''
+
+[tool.flake8]
+max-line-length = 130
+extend-ignore = ["D203", "E203", "E251", "E266", "E302", "E305", "E401", "E402", "E501", "F401", "F403", "W503"]
+exclude = [".git", "__pycache__", "dist"]
+max-complexity = 10
+
+[tool.isort]
+atomic = true
+profile = "black"
+line_length = 130
+skip_gitignore = true
+
+[tool.ruff]
+# Exclude a variety of commonly ignored directories.
+exclude = [
+    ".bzr",
+    ".direnv",
+    ".eggs",
+    ".git",
+    ".git-rewrite",
+    ".hg",
+    ".ipynb_checkpoints",
+    ".mypy_cache",
+    ".nox",
+    ".pants.d",
+    ".pyenv",
+    ".pytest_cache",
+    ".pytype",
+    ".ruff_cache",
+    ".svn",
+    ".tox",
+    ".venv",
+    ".vscode",
+    "__pypackages__",
+    "_build",
+    "buck-out",
+    "build",
+    "dist",
+    "node_modules",
+    "site-packages",
+    "venv",
+]
+
+# Same as Black.
+line-length = 130
+indent-width = 4
+target-version = "py310"
+
+[tool.ruff.lint]
+# Enable Pyflakes (`F`) and a subset of the pycodestyle (`E`)  codes by default.
+# Unlike Flake8, Ruff doesn't enable pycodestyle warnings (`W`) or
+# McCabe complexity (`C901`) by default.
+select = ["E4", "E7", "E9", "F"]
+ignore = []
+
+# Allow fix for all enabled rules (when `--fix`) is provided.
+fixable = ["ALL"]
+unfixable = []
+
+# Allow unused variables when underscore-prefixed.
+dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
+
+[tool.ruff.format]
+# Like Black, use double quotes for strings.
+quote-style = "double"
+
+# Like Black, indent with spaces, rather than tabs.
+indent-style = "space"
+
+# Like Black, respect magic trailing commas.
+skip-magic-trailing-comma = false
+
+# Like Black, automatically detect the appropriate line ending.
+line-ending = "auto"
+
+# Enable auto-formatting of code examples in docstrings. Markdown,
+# reStructuredText code/literal blocks and doctests are all supported.
+#
+# This is currently disabled by default, but it is planned for this
+# to be opt-out in the future.
+docstring-code-format = false
+
+# Set the line length limit used when formatting code snippets in
+# docstrings.
+#
+# This only has an effect when the `docstring-code-format` setting is
+# enabled.
+docstring-code-line-length = "dynamic"
+
+[tool.poetry.dependencies]
+python = ">=3.9 <3.11"
+transformers = "^4.42.4"
+mathgenerator = {git = "https://github.com/synapse-alpha/mathgenerator.git", rev = "main"}
+sympy = "^1.13.0"
+tqdm = "^4.66.4"
+requests = "^2.32.3"
+numpy = "1.26.4"
+bittensor = "^7.3.0"
```

Comment:
Have you tested `7.3.0`? I remember @bkb2135 mentioned there were some issues with higher bittensor versions

Also it's better to use exact versions for all the packages, rather than `^` or `>=` to avoid getting into breaking changes issues

Reply:
This is one of the neat things about poetry! In the poetry.lock file we have exact versions specified (so exact versions are being installed & we won't be affected by breaking changes) but in pyproject.toml you generally want to specify ranges so that we are able to update to new versions easily in the future (you can simply run `poetry update` and it'll move to the newest versions based on the pyproject.toml and save that in the .lock file).

So it allows us to intentionally add a contraint when we want and have the freedom to upgrade at our leisure whenever we want. But current versions are already locked, so if bittensor upgrades we'll still be installing whatever it says currently in poetry.lock.

That's actually one of the nice advantages of poetry over pip:)

Reply:
Makes sense, in that case we should consider committing `poetry.lock`